src/java/org/apache/cassandra/service/accord/AccordTopology.java

```diff
@@ -134,22 +138,34 @@ private static KeyspaceShard forRange(KeyspaceMetadata keyspace, Range<Token> ra
                                                  .map(directory::peerId)
                                                  .map(AccordTopology::tcmIdToAccord)
                                                  .collect(Collectors.toSet());
+            Set<Node.Id> leaving = allLeaving.isEmpty() ?
+                                   Collections.emptySet() :
+                                   allLeaving.stream()
+                                             .filter(e -> writeEndpoints.contains(e))
+                                             .map(directory::peerId)
+                                             .map(AccordTopology::tcmIdToAccord)
+                                             .collect(Collectors.toSet());
 
-            return new KeyspaceShard(keyspace, range, nodes, pending);
+            return new KeyspaceShard(keyspace, range, nodes, pending, leaving);
         }
 
-        public static List<KeyspaceShard> forKeyspace(KeyspaceMetadata keyspace, DataPlacements placements, Directory directory, ShardLookup lookup)
+        public static List<KeyspaceShard> forKeyspace(KeyspaceMetadata keyspace,
+                                                      Function<KeyspaceMetadata, DataPlacement> placements,
+                                                      Function<KeyspaceMetadata, DataPlacement> settledPlacements,
+                                                      Directory directory, ShardLookup lookup)
         {
-            ReplicationParams replication = keyspace.params.replication;
-            DataPlacement placement = placements.get(replication);
+//            ReplicationParams replication = keyspace.params.replication;
+            DataPlacement placement = placements.apply(keyspace);
+            DataPlacement settled = settledPlacements.apply(keyspace);
+            Set<InetAddressAndPort> leaving = Sets.difference(placement.writes.byEndpoint().keySet(), settled.writes.byEndpoint().keySet());
```

Comment:
I should cleanup this logic, we are already using the `KeyspaceMetadata` so we really don't need functions, we can just pass in the 2

Reply:
FWIW, we could integrate more closely and have `AccordTopology` inside TCM. I somehow thought this also was a plan at some point.

Reply:
I personally prefer the separation... we support the following yaml `accord.enabled: false`, so we won't create `Topology` when disabled, but if we link in TCM we would and would need to also handle that in the `ClusterMetadata`?